presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java

```diff
@@ -835,54 +872,90 @@ private void queryFinished(ManagedQueryExecution query)
             }
-            else {
-                for (Iterator<InternalResourceGroup> iterator = dirtySubGroups.iterator(); iterator.hasNext(); ) {
-                    InternalResourceGroup subGroup = iterator.next();
-                    long oldMemoryUsageBytes = subGroup.cachedMemoryUsageBytes;
-                    cachedMemoryUsageBytes -= oldMemoryUsageBytes;
-                    subGroup.internalRefreshStats();
-                    cachedMemoryUsageBytes += subGroup.cachedMemoryUsageBytes;
-                    if (!subGroup.isDirty()) {
-                        iterator.remove();
-                    }
-                    if (oldMemoryUsageBytes != subGroup.cachedMemoryUsageBytes || subGroup.isDirty.get()) {
-                        subGroup.updateEligibility();
-                        subGroup.isDirty.set(false);
-                    }
+            waitingQueuedQueries.set(waitingQueuedQueryCount);
+        }
+    }
+
+    // Memory usage stats are expensive to maintain, so this method must be called periodically to update them
+    private void updateMemoryUsageInfo()
+    {
+        checkState(Thread.holdsLock(root), "Must hold lock to update the memory usage info");
+        if (subGroups.isEmpty()) {
+            cachedMemoryUsageBytes = 0;
+            for (ManagedQueryExecution query : runningQueries) {
+                cachedMemoryUsageBytes += query.getUserMemoryReservation().toBytes();
+            }
+            Optional<ResourceGroupRuntimeInfo> resourceGroupRuntimeInfo = getAdditionalRuntimeInfo();
+            resourceGroupRuntimeInfo.ifPresent(groupRuntimeInfo -> cachedMemoryUsageBytes += groupRuntimeInfo.getMemoryUsageBytes());
+        }
+        else {
+            for (Iterator<InternalResourceGroup> iterator = dirtySubGroups.iterator(); iterator.hasNext(); ) {
+                InternalResourceGroup subGroup = iterator.next();
+                long oldMemoryUsageBytes = subGroup.cachedMemoryUsageBytes;
+                cachedMemoryUsageBytes -= oldMemoryUsageBytes;
+                subGroup.updateMemoryUsageInfo();
+                cachedMemoryUsageBytes += subGroup.cachedMemoryUsageBytes;
+                if (!subGroup.isDirty()) {
+                    iterator.remove();
+                }
+                if (oldMemoryUsageBytes != subGroup.cachedMemoryUsageBytes || subGroup.isDirty.get()) {
+                    subGroup.updateEligibility();
+                    subGroup.isDirty.set(false);
                 }
             }
         }
     }
 
+    @VisibleForTesting
+    void refreshStats()
+    {
+        synchronized (root) {
+            internalRefreshStats();
+        }
+    }
+
+    protected void internalRefreshStats()
+    {
+        checkState(Thread.holdsLock(root), "Must hold lock to refresh stats");
+        synchronized (root) {
+            updateMemoryUsageInfo();
+            updateQueriesQueuedOnInternal();
+            updateWaitingQueuedQueries();
```

Comment:
`internalRefreshStats` called for during the `refreshAndStartQueries` loop. Calculating queued queries for all those loop is expensive and can delay queries to start from the queue. Better to have it done at a regular internal in a seperate thread instead of part of `internalRefreshStats`